dht.go

```diff
@@ -498,6 +526,46 @@ func (dht *IpfsDHT) SetMode(m DHTMode) error {
 	}
 }
 
+func (dht *IpfsDHT) dynamicModeSwitching(ctx context.Context) goprocess.Process {
+	proc := goprocessctx.WithContext(ctx)
+	watch := func(proc goprocess.Process) {
+		var (
+			target   DHTMode
+			debounce <-chan time.Time
+		)
+		for {
+			select {
+			case ev := <-dht.subscriptions.evtLocalRoutability.Out():
+				switch ev.(type) {
+				case event.EvtLocalRoutabilityPrivate, event.EvtLocalRoutabilityUnknown:
+					target = ModeClient
+				case event.EvtLocalRoutabilityPublic:
+					target = ModeServer
+				}
+				if debounce == nil {
+					debounce = time.After(DynamicModeSwitchDebouncePeriod)
```

Comment:
This is a throttle. That works but a debounce would probably be better. To do that, we'd:

1. Have a `time.Timer`.
2. Call `myTimer.Reset(deadline)` every time we receive an event.

Reply:
And if we have a debounce set, and we switch back to the mode we were previously in, just cancel the operation and clear out the timer.

Reply:
Actually debouncing now ;-) https://github.com/libp2p/go-libp2p-kad-dht/pull/367/commits/87f6b0c908de352e67034476588d2bc67171aa60